File: packages-gen/js/event/util-task.js
const xtd = require("@xtalk/lang/common-data.js")

const xtst = require("@xtalk/lang/common-sort-topo.js")

function new_task(id,deps,args,opts){
  return Object.assign({"::":"loader.task","id":id,"deps":deps,"args":args},opts);
}

function task_load(task){
  let {args,assert_fn,check_fn,get_fn,id,load_fn,load_no_check} = task;
  if("function" == (typeof assert_fn)){
    if(!assert_fn()){
      throw "Assertion Failed - " + id;
    }
  }
  if(true != load_no_check){
    let curr = ("function" == (typeof get_fn)) ? get_fn() : null;
    let check = ("function" == (typeof check_fn)) ? check_fn(curr) : null;
    if(true == check){
      return Promise.resolve().then(function (){
        return curr;
      });
    }
  }
  if("function" == (typeof args)){
    args = args();
  }
  args = ((null == args) ? [] : args);
  return Promise.resolve().then(function (){
    return load_fn.apply(null,args);
  });
}

function task_unload(task){
  let {check_fn,get_fn,unload_fn,unload_no_check} = task;
  if(true != unload_no_check){
    let curr = ("function" == (typeof get_fn)) ? get_fn() : null;
    let check = ("function" == (typeof check_fn)) ? check_fn(curr) : null;
    if(true != check){
      return Promise.resolve().then(function (){
        return false;
      });
    }
  }
  return Promise.resolve().then(function (){
    return unload_fn();
  }).then(function (_){
    return true;
  });
}

function unload_tasks_loop(tasks,completed,ids,out,hook_fn){
  if(0 == ids.length){
    return Promise.resolve().then(function (){
      return out;
    });
  }
  let id = ids[0];
  let rest = ids.slice(1,ids.length);
  if(true != completed[id]){
    return unload_tasks_loop(tasks,completed,rest,out,hook_fn);
  }
  delete(completed[id]);
  let task = tasks[id];
  return task_unload(task).then(function (unloaded){
    if(null != hook_fn){
      hook_fn(id,unloaded);
    }
    out.push([id,unloaded]);
    return unload_tasks_loop(tasks,completed,rest,out,hook_fn);
  });
}

function new_loader_blank(){
  return {
    "::":"loader",
    "completed":{},
    "loading":{},
    "errored":null,
    "order":[],
    "tasks":{}
  };
}

function add_tasks(loader,tasks){
  let prev = loader["tasks"];
  let all = xtd.arr_assign(Object.values(prev),tasks);
  let deps = all.map(function (e){
    return [e["id"],e["deps"]];
  });
  return Object.assign(loader,{
    "order":xtst.sort_topo(deps),
    "tasks":xtd.arr_juxt(all,function (e){
        return e["id"];
      },xtd.clone_nested)
  });
}

function new_loader(tasks){
  return add_tasks(new_loader_blank(),tasks);
}

function list_loading(loader){
  return Object.keys(loader["loading"]);
}

function list_completed(loader){
  return Object.keys(loader["completed"]);
}

function list_incomplete(loader){
  let {completed,order,tasks} = loader;
  let out = [];
  for(let id of order){
    if((null != tasks[id]) && (true != completed[id])){
      out.push(id);
    }
  };
  return out;
}

function list_waiting(loader){
  let {completed,loading,order,tasks} = loader;
  let out = [];
  for(let id of order){
    let task = tasks[id];
    if((null != task) && (true != loading[id]) && (true != completed[id]) && task["deps"].every(function (dep_id){
      return true == completed[dep_id];
    })){
      out.push(id);
    }
  };
  return out;
}

function load_tasks_single(loader,id,hook_fn,complete_fn,loop_fn){
  let {completed,loading,tasks} = loader;
  let task = tasks[id];
  loading[id] = true;
  return task_load(task).catch(function (err){
    delete(loading[id]);
    loader["errored"] = id;
    if(null != hook_fn){
      hook_fn(id,false);
    }
    if(null != complete_fn){
      complete_fn(err);
    }
    throw err;
  }).then(function (res){
    delete(loading[id]);
    completed[id] = true;
    if(null != hook_fn){
      hook_fn(id,true);
    }
    return (null != loop_fn) ? loop_fn(loader,hook_fn,complete_fn) : res;
  });
}

function load_tasks(loader,hook_fn,complete_fn){
  let {errored,order} = loader;
  if(null != errored){
    let reject_fn = function (){
      throw "ERR - Task Errored - " + errored;
    };
    return Promise.resolve().then(reject_fn);
  }
  let waiting = list_waiting(loader);
  if(0 < waiting.length){
    let waiting_lu = xtd.arr_juxt(waiting,function (id){
      return id;
    },function (id){
      return true;
    });
    let next_id =     (xtd.arr_keep(order,function (id){
          if(true == waiting_lu[id]){
            return id;
          }
        }))[0];
    return load_tasks_single(loader,next_id,hook_fn,complete_fn,load_tasks);
  }
  let incomplete = list_incomplete(loader);
  if(0 == incomplete.length){
    if(null != complete_fn){
      complete_fn(true);
    }
    return Promise.resolve().then(function (){
      return loader;
    });
  }
  return Promise.resolve().then(function (){
    return loader;
  });
}

function unload_tasks(loader,hook_fn){
  let {completed,order,tasks} = loader;
  let rorder = order.slice().reverse();
  return unload_tasks_loop(tasks,completed,rorder,[],hook_fn);
}

module.exports = {
  ["new_task"]:new_task,
  ["task_load"]:task_load,
  ["task_unload"]:task_unload,
  ["unload_tasks_loop"]:unload_tasks_loop,
  ["new_loader_blank"]:new_loader_blank,
  ["add_tasks"]:add_tasks,
  ["new_loader"]:new_loader,
  ["list_loading"]:list_loading,
  ["list_completed"]:list_completed,
  ["list_incomplete"]:list_incomplete,
  ["list_waiting"]:list_waiting,
  ["load_tasks_single"]:load_tasks_single,
  ["load_tasks"]:load_tasks,
  ["unload_tasks"]:unload_tasks
}

File: packages-gen/dart/libs/xt-event/lib/util-task.dart
import 'package:xtalk_lang/common-data.dart' as xtd;
import 'package:xtalk_lang/common-sort-topo.dart' as xtst;
import 'dart:async';



new_task(id, deps, args, opts) {
  return xtd.obj_assign(
    <dynamic, dynamic>{"::":"loader.task","id":id,"deps":deps,"args":args},
    opts
  );
}

task_load(task) {
  var args = task["args"];
  var assert_fn = task["assert_fn"];
  var check_fn = task["check_fn"];
  var get_fn = task["get_fn"];
  var id = task["id"];
  var load_fn = task["load_fn"];
  var load_no_check = task["load_no_check"];
  if((assert_fn.runtimeType).toString().contains("Function") || (assert_fn.runtimeType).toString().contains("=>") || (assert_fn).toString().startsWith("Closure")){
    if(!(() {
      var dart_truthy__50840 = Function.apply((assert_fn as Function),<dynamic>[]);
      return (null != dart_truthy__50840) && (false != dart_truthy__50840);
    })()){
      throw "Assertion Failed - " + id;
    }
  }
  if(true != load_no_check){
    var curr = ((get_fn.runtimeType).toString().contains("Function") || (get_fn.runtimeType).toString().contains("=>") || (get_fn).toString().startsWith("Closure")) ? Function.apply((get_fn as Function),<dynamic>[]) : null;
    var check = ((check_fn.runtimeType).toString().contains("Function") || (check_fn.runtimeType).toString().contains("=>") || (check_fn).toString().startsWith("Closure")) ? Function.apply((check_fn as Function),<dynamic>[curr]) : null;
    if(true == check){
      return Future.sync(() {
        return curr;
      });
    }
  }
  if((args.runtimeType).toString().contains("Function") || (args.runtimeType).toString().contains("=>") || (args).toString().startsWith("Closure")){
    args = args();
  }
  args = ((null == args) ? <dynamic>[] : args);
  return Future.sync(() {
    return Function.apply(load_fn,args);
  });
}

task_unload(task) {
  var check_fn = task["check_fn"];
  var get_fn = task["get_fn"];
  var unload_fn = task["unload_fn"];
  var unload_no_check = task["unload_no_check"];
  if(true != unload_no_check){
    var curr = ((get_fn.runtimeType).toString().contains("Function") || (get_fn.runtimeType).toString().contains("=>") || (get_fn).toString().startsWith("Closure")) ? Function.apply((get_fn as Function),<dynamic>[]) : null;
    var check = ((check_fn.runtimeType).toString().contains("Function") || (check_fn.runtimeType).toString().contains("=>") || (check_fn).toString().startsWith("Closure")) ? Function.apply((check_fn as Function),<dynamic>[curr]) : null;
    if(true != check){
      return Future.sync(() {
        return false;
      });
    }
  }
  return ((Future.sync(() => Future.sync(() {
    return Function.apply((unload_fn as Function),<dynamic>[]);
  }))) as Future<dynamic>).then((value) async { return await Function.apply((_) {
    return true;
  },<dynamic>[value]); });
}

unload_tasks_loop(tasks, completed, ids, out, hook_fn) {
  if(0 == ids.length){
    return Future.sync(() {
      return out;
    });
  }
  var id = ids[0];
  var rest = ids.sublist(1 - 0,ids.length);
  if(true != completed[id]){
    return unload_tasks_loop(tasks,completed,rest,out,hook_fn);
  }
  completed.remove(id);
  var task = tasks[id];
  return ((Future.sync(() => task_unload(task))) as Future<dynamic>).then((value) async { return await Function.apply((unloaded) {
    if(null != hook_fn){
      Function.apply((hook_fn as Function),<dynamic>[id,unloaded]);
    }
    out.add(<dynamic>[id,unloaded]);
    return unload_tasks_loop(tasks,completed,rest,out,hook_fn);
  },<dynamic>[value]); });
}

new_loader_blank() {
  return <dynamic, dynamic>{
    "::":"loader",
    "completed":<dynamic, dynamic>{},
    "loading":<dynamic, dynamic>{},
    "errored":null,
    "order":<dynamic>[],
    "tasks":<dynamic, dynamic>{}
  };
}

add_tasks(loader, tasks) {
  var prev = loader["tasks"];
  var all = xtd.arr_assign(List<dynamic>.from(( prev ).values),tasks);
  var deps = xtd.arr_map(all,(e) {
    return <dynamic>[e["id"],e["deps"]];
  });
  return xtd.obj_assign(loader,<dynamic, dynamic>{
    "order":xtst.sort_topo(deps),
    "tasks":xtd.arr_juxt(all,(e) {
        return e["id"];
      },xtd.clone_nested)
  });
}

new_loader(tasks) {
  return add_tasks(new_loader_blank(),tasks);
}

list_loading(loader) {
  return List<dynamic>.from(( loader["loading"] ).keys);
}

list_completed(loader) {
  return List<dynamic>.from(( loader["completed"] ).keys);
}

list_incomplete(loader) {
  var completed = loader["completed"];
  var order = loader["order"];
  var tasks = loader["tasks"];
  var out = <dynamic>[];
  var arr_50847 = order;
  for(var i50848 = 0; i50848 < arr_50847.length; ++i50848){
    var id = arr_50847[i50848];
    if((null != tasks[id]) && (true != completed[id])){
      out.add(id);
    }
  };
  return out;
}

list_waiting(loader) {
  var completed = loader["completed"];
  var loading = loader["loading"];
  var order = loader["order"];
  var tasks = loader["tasks"];
  var out = <dynamic>[];
  var arr_50869 = order;
  for(var i50870 = 0; i50870 < arr_50869.length; ++i50870){
    var id = arr_50869[i50870];
    var task = tasks[id];
    if((null != task) && (true != loading[id]) && (true != completed[id]) && (() {
      var dart_truthy__50839 = xtd.arr_every(task["deps"],(dep_id) {
        return true == completed[dep_id];
      });
      return (null != dart_truthy__50839) && (false != dart_truthy__50839);
    })()){
      out.add(id);
    }
  };
  return out;
}

load_tasks_single(loader, id, hook_fn, complete_fn, loop_fn) {
  var completed = loader["completed"];
  var loading = loader["loading"];
  var tasks = loader["tasks"];
  var task = tasks[id];
  loading[id] = true;
  return ((Future.sync(() => (() async { try { return await ((Future.sync(() => task_load(task))) as Future<dynamic>); } catch (err) { return await Future.sync(() => Function.apply((err) {
    loading.remove(id);
    loader["errored"] = id;
    if(null != hook_fn){
      Function.apply((hook_fn as Function),<dynamic>[id,false]);
    }
    if(null != complete_fn){
      Function.apply((complete_fn as Function),<dynamic>[err]);
    }
    throw err;
  },<dynamic>[err])); } })())) as Future<dynamic>).then((value) async { return await Function.apply((res) {
    loading.remove(id);
    completed[id] = true;
    if(null != hook_fn){
      Function.apply((hook_fn as Function),<dynamic>[id,true]);
    }
    return (null != loop_fn) ? Function.apply((loop_fn as Function),<dynamic>[loader,hook_fn,complete_fn]) : res;
  },<dynamic>[value]); });
}

load_tasks(loader, hook_fn, complete_fn) {
  var errored = loader["errored"];
  var order = loader["order"];
  if(null != errored){
    var reject_fn = () {
      throw "ERR - Task Errored - " + errored;
    };
    return Future.sync(reject_fn);
  }
  var waiting = list_waiting(loader);
  if(0 < waiting.length){
    var waiting_lu = xtd.arr_juxt(waiting,(id) {
      return id;
    },(id) {
      return true;
    });
    var next_id =     (xtd.arr_keep(order,(id) {
          if(true == waiting_lu[id]){
            return id;
          }
        }))[0];
    return load_tasks_single(loader,next_id,hook_fn,complete_fn,load_tasks);
  }
  var incomplete = list_incomplete(loader);
  if(0 == incomplete.length){
    if(null != complete_fn){
      Function.apply((complete_fn as Function),<dynamic>[true]);
    }
    return Future.sync(() {
      return loader;
    });
  }
  return Future.sync(() {
    return loader;
  });
}

unload_tasks(loader, hook_fn) {
  var completed = loader["completed"];
  var order = loader["order"];
  var tasks = loader["tasks"];
  var rorder = xtd.arr_reverse(order);
  return unload_tasks_loop(tasks,completed,rorder,<dynamic>[],hook_fn);
}